README.md
![build status](https://github.com/flu-crew/flutile/actions/workflows/python-app.yml/badge.svg)
![PyPI](https://img.shields.io/pypi/v/flutile.svg)

# flutile

## Installation

``` sh
pip install flutile
```

## Commands

### `aadiff`

`aadiff` takes a multiple-sequence alignment as an input and creates a
character difference table. This command is designed for preparing amino acid
difference tables. Below is an example of a comparison of 4 H1 sequences.

`flutile aadiff --subtype=H1 mufile.faa`

| site  | A02479030 | SD0246 | SD0272 | SD0136 |
| ----  | --------- | ------ | -------| ------ |
| -3    | A         |        | T      |        |
| -2    | N         |        | S      |        |
| -1    | A         | T      |        |        |
| 1     | D         |        |        |        |
| 154   | K         |        |        |        |
| 154+1 | -         |        |        | X      |
| 154+2 | -         |        |        | X      |
| 155   | D         | S      |        | X      |
| 156   | D         | G      | N      | X      |

The `--subtype=H1` argument tells `flutile` to align the inputs against an H1
reference (A/United Kingdom/1/1933). The reference is used to determine
relative indices (the `sites` column). The index reference is used only for
indexing and does not appear in the final table. The first three rows (sites
-3, -2, -1) align to the three residues at the end of the signal peptide. Site
1 is the first residue in the mature peptide. Any gaps in the reference
alignment are indexed as `<ref_id>+<offset>`, for example 154+1 and 154+2 are
positions 1 and 2 residues after the reference position 154.

`flutile` uses the references from (Burke 2014):

--- | ------------------------------------------------- | ------------------- | ---------------------- |
H1  | A/United Kingdom/1/1933                           | MKARLLVLLCALAATDA   | DTICIGYHANNS           |
H2  | A/Singapore/1/1957                                | MAIIYLILLFTAVRG     | DQICIGYHANNS           |
H3  | A/Aichi/2/1968                                    | MKTIIALSYIFCLPLG    | QDLPGNDNSTATLCLGHHAVPN |
H4  | A/swine/Ontario/01911–2/1999                      | MLSIAILFLLIAEGSS    | QNYTGNPVICLGHHAVSN     |
H5  | A/Vietnam/1203/2004                               | MEKIVLLFAIVSLVKS    | DQICIGYHANNS           |
H6  | A/chicken/Taiwan/0705/1999                        | MIAIIVIATLAAAGKS    | DKICIGYHANNS           |
H7  | A/Netherlands/219/2003                            | MNTQILVFALVASIPTNA  | DKICLGHHAVSN           |
H8  | A/turkey/Ontario/6118/1968                        | MEKFIAIAMLLASTNA    | YDRICIGYQSNNS          |
H9  | A/swine/Hong Kong/9/1998                          | MEAASLITILLVVTASNA  | DKICIGYQSTNS           |
H10 | A/mallard/bavaria/3/2006                          | MYKIVVIIALLGAVKG    | LDKICLGHHAVAN          |
H11 | A/duck/England/1/1956                             | MEKTLLFAAIFLCVKA    | DEICIGYLSNNS           |
H12 | A/duck/Alberta/60/1976                            | MEKFIILSTVLAASFA    | YDKICIGYQTNNS          |
H13 | A/gull/Maryland/704/1977                          | MALNVIATLTLISVCVHA  | DRICVGYLSTNS           |
H14 | A/mallard/Astrakhan/263/1982                      | MIALILVALALSHTAYS   | QITNGTTGNPIICLGHHAVEN  |
H15 | A/duck/Australia/341/1983                         | MNTQIIVILVLGLSMVRS  | DKICLGHHAVAN           |
H16 | A/black-headed-gull/Turkmenistan/13/1976          | MMIKVLYFLIIVLGRYSKA | DKICIGYLSNNS           |
H17 | A/little-yellow-shouldered bat/Guatemala/060/2010 | MELIILLILLNPYTFVLG  | DRICIGYQANQN           |
H18 | A/flat-faced bat/Peru/033/2010                    | MITILILVLPIVVG      | DQICIGYHSNNS           |

 * The H3 signal peptide appears to actually be `MKTIIALSYIFCLALG`

### `represent`

`represent` takes a multiple-sequence alignment as input and removes entries
that are similar in sequence and time. The function requires that headers have
a date term (with format year/month/day). For example:

```
>A|1990-01-02
GATACA
>B|1990-02-02
CATATA
```

There may be gaps in the alignment. Sequences in the alignment that are
separated by less than or equal to `--max-day-sep` and that are a sequence
identity of greather than or equal to `--min-pident-sep` will be clustered
together. A single representative is sampled from each cluster (the latest one
with ties resolved by order).


### `trim`

Extract the HA1 regions from (currently) H1 and H3 HA proteins.

### References

  1. Burke, Smith (2014) *A Recommended Numbering Scheme for Influenza A HA Subtypes*


ChangeLog.md
1.0.1 [2022-02-16]
------------------

 * Fix dependency handling in setup

1.0.0 [2022-02-09]
------------------

 My final commit as an ORISE postdoc

0.13.5 [2022-01-11]
-------------------

 * Export mypy types

0.13.4 [2021-12-30]
-------------------

 * Fix compatibility bug with octofludb

0.13.3 [2021-12-30]
-------------------

 * Add verbose option, by default do not print mafft messages
 * Make --annotation-table argument take a single filename (rather than
   comma-delimited) and allow the option to be used multiple times. E.g.:
        flutile aadiff --annotation-table foo.txt --annotation-table bar.txt
 * Migrate from Travis to GitHub Actions for CI
 * Add MyPy type annotations
 * Removed mixed data from test files

0.13.2 [2021-12-16]
-------------------

 * Add version command

0.13.1 [2021-11-18]
-------------------

 * Fix edge cases
 * Pass filehandles to functions that print

0.13.0 [2021-10-03]
-------------------

 * Set caton82 antigenic sites exactly to the sites described in paper

0.12.3 [2021-07-xx]
-------------------

 * Fix flake8 issues

0.12.2 [2021-07-20]
-------------------

 * Fix HA1 trimming for all subtypes
 * Test agreement of flutile predicted HA1 with 50 randomly selected strains
   for each subtype from genbank

0.12.1 [2021-07-15]
-------------------

 * Fix bug in dna2dna conversions and add tests

0.12.0 [2021-05-05]
-------------------

 * Don't spam .tmp temporary alignment files everywhere
 * Add `--count` flag to aggregate counts in aadiff tables

0.11.0 [2021-04-12]
-------------------

 * Add support for for NA subtypes and internal genes 


0.10.0 [2021-03-24]
-------------------

 * Extract motifs by parsing interval expressions such as:
   - "motif=34,56-60"
   - "34"
   - "162,166,167-190"
 * Raise a nice error message when `--subtype` is missing 

0.9.0 [2021-03-19]
------------------

 * Add `trim motif` command for extracting motifs

0.8.0 [2021-03-19]
------------------

 * Fix type in Caton82 annotations
 * Generalize trim ha1 to work for all 18 HA subtypes 

0.7.0 [2020-12-17]
------------------

`annotate` subcommand for mapping between subtypes

0.6.1 [2020-12-03]
------------------

Add wiley81 H3 reference

0.6.0 [2020-12-03]
------------------

allow aadiff annotations 

   - `--caton82` for antigenic sites
   - `--annotation-tables=STR` for list of annotation tables to load
   - `--join-annotations` to join all annotations into one column

0.5.0 [2020-09-14]
------------------

 * Fix bug in h3 trim
 * Add tests for trim
 * Add sequence samples from all h1 and h3 clades
 * Add aadiff subcommand with absolute HA indexing against subtype refs

0.4.0 [2020-09-11]

Add `trim`

0.3.2 [2020-08-13]

Fix bug in --print-groups

0.3.1 [2020-08-12]

Bug fixes

0.3.0 [2020-08-06]
------------------

Changes to represent:
 * Add --print-groups option
 * Add progress bar
 * Only consider dates if --max-day-sep argument is given (no default)
 * If --max-day-sep is given, keep the latest entry from each group, otherwise
   keep first alphabetically
 * Do not parse states unless --same-state is given 

0.2.0 [2020-01-06]
------------------

 * add group by USA state

0.1.0 [2019-12-20]
------------------

 * initial release


flutile/ui.py
from __future__ import annotations
import click
import sys
import flutile.functions as fun
from flutile.parameters import (
    AadiffOpts,
    MafftOpts,
    SeqOpts,
    parse_conversion,
)
from flutile.version import __version__
from typing import TextIO, List, Optional

INT_SENTINEL = 1e9

CONTEXT_SETTINGS = dict(help_option_names=["-h", "--help"])

mafft_exe_opt = click.option(
    "--mafft-exe",
    default="mafft",
    type=str,
    help="Path to MAFFT alignment tool executable",
)

ha_subtypes = [
    "H1",
    "H2",
    "H3",
    "H4",
    "H5",
    "H6",
    "H7",
    "H8",
    "H9",
    "H10",
    "H11",
    "H12",
    "H13",
    "H14",
    "H15",
    "H16",
    "H17",
    "H18",
]

na_subtypes = [
    "N1",
    "N2",
    "N3",
    "N4",
    "N5",
    "N6",
    "N7",
    "N8",
    "N9",
    "N10",
    "N11",
]

internal_subtypes = [
    "PB2",
    "PB1",
    "PA",
    "NP",
    "M1",
    "M",
    "MP",
    "NS1",
    "NS",
]

ha_subtype_opt = click.option(
    "--subtype",
    type=click.Choice(
        ha_subtypes,
        case_sensitive=False,
    ),
    required=True,
    help="The HA segment subtype of all entries in the input fasta file. The numbering will be relative to the start of the mature peptide, with offsets as defined by (Burke 2014).",
)

verbose_opt = click.option(
    "--verbose",
    is_flag=True,
    default=False,
    help="Print verbose log messages",
)

subtype_opt = click.option(
    "--subtype",
    type=click.Choice(
        ha_subtypes + na_subtypes + internal_subtypes,
        case_sensitive=False,
    ),
    required=True,
    help="The segment subtype of all entries in the input fasta file. For HA subtypes, the numbering will be relative to the start of the mature peptide, with offsets as defined by (Burke 2014). When the `--keep-signal` flag is set for an HA subtype, and for all NA and internal gene segments, indexing is from the initial methionine. For NA, you may use N1 to index by conventional WNS numbering.",
)

ha_subtype_no_keep_opt = click.option(
    "--subtype",
    type=click.Choice(
        ha_subtypes,
        case_sensitive=False,
    ),
    required=True,
    help="The segment subtype of all strains in the input fasta file",
)

subtype_no_keep_opt = click.option(
    "--subtype",
    type=click.Choice(
        ha_subtypes + na_subtypes + internal_subtypes,
        case_sensitive=False,
    ),
    required=True,
    help="The segment subtype of all strains in the input fasta file",
)

as_fasta_opt = click.option(
    "--fasta",
    is_flag=True,
    default=False,
    help="Return motif as a FASTA file with all motifs concatenated (by default a table is printed)",
)

make_consensus_opt = click.option(
    "--make-consensus", is_flag=True, help="Add a sequence consensus column"
)

consensus_as_reference_opt = click.option(
    "--consensus-as-reference",
    is_flag=True,
    help="Use the consensus as the reference column",
)

keep_signal_opt = click.option(
    "--keep-signal",
    is_flag=True,
    help="Number relative to the initial Methionine (do not trim the signal peptide)",
)

count_variants_opt = click.option(
    "--count",
    is_flag=True,
    help="Instead of listing each strain in the alignment, count the amino acid variations",
)

annotation_tables_opt = click.option(
    "--annotation-tables",
    multiple=True,
    help="A filename containing annotations. The first column must contain relative indices. These may be negative to refer to positions before the start of the HA1 region or may refer to indels (e.g., 42+1 for an insertion after the 42 site in the reference). The table is expected to have a header with column names.",
)

join_annotations_opt = click.option(
    "--join-annotations",
    help="Join all annotation columns, seperating values with commas",
    is_flag=True,
)

caton82_opt = click.option(
    "--caton82", is_flag=True, help="Add H1 antigenic sites from [Caton 1982]"
)

wiley81_opt = click.option(
    "--wiley81", is_flag=True, help="Add H3 antigenic sites from [Wiley 1981]"
)

motif_range_opt = click.option(
    "--motif",
    "-m",
    type=str,
    multiple=True,
    help="A string describing the range of a motif. It can be an integer ('162'), a range ('162-169'), or a composite ('162,165-169'). The motif can also be given a name ('bob=162,165-169').",
)


@click.command(
    name="aadiff",
    help="Compare differences between sequences. The input fasta file does NOT need to be aligned. If a subtype is specified, indexing will be relative to the index reference (Burke 2014). If no subtype is specified, the numbering will be relative to the first entry in fasta file (or the consensus sequence if --consensus-as-reference is set)",
)
@click.argument("faa", default=sys.stdin, type=click.File())
@subtype_opt
@make_consensus_opt
@consensus_as_reference_opt
@keep_signal_opt
@count_variants_opt
@mafft_exe_opt
@verbose_opt
@annotation_tables_opt
@join_annotations_opt
@caton82_opt
@wiley81_opt
def aadiff_cmd(
    faa: TextIO,
    subtype: str,
    make_consensus: bool,
    consensus_as_reference: bool,
    keep_signal: bool,
    count: bool,
    mafft_exe: str,
    verbose: bool,
    annotation_tables: List[str],
    join_annotations: bool,
    caton82: bool,
    wiley81: bool,
):
    mafft_opts = MafftOpts(mafft_exe=mafft_exe, verbose=verbose)
    seq_opts = SeqOpts(subtype=subtype, keep_signal=keep_signal)
    aadiff_opts = AadiffOpts(
        make_consensus=make_consensus,
        consensus_as_reference=consensus_as_reference,
        remove_unchanged=True,
        annotation_tables=annotation_tables,
        join_annotations=join_annotations,
        count=count,
        caton82=caton82,
        wiley81=wiley81,
    )
    for row in fun.referenced_aadiff_table(
        faa, mafft_opts=mafft_opts, seq_opts=seq_opts, aadiff_opts=aadiff_opts
    ):
        print("\t".join(row))


@click.command(
    name="annotate",
    help="Tabulate differences between HA sequences. This command is like aadiff except it adds columns mapping to indices across subtypes (using Burke 2014 numbering).",
)
@click.argument("faa", default=sys.stdin, type=click.File())
@ha_subtype_opt
@make_consensus_opt
@consensus_as_reference_opt
@mafft_exe_opt
@verbose_opt
@annotation_tables_opt
@join_annotations_opt
@caton82_opt
@wiley81_opt
def annotate_cmd(
    faa: TextIO,
    subtype: str,
    make_consensus: bool,
    consensus_as_reference: bool,
    mafft_exe: str,
    verbose: bool,
    annotation_tables: List[str],
    join_annotations: bool,
    caton82: bool,
    wiley81: bool,
):
    mafft_opts = MafftOpts(mafft_exe=mafft_exe, verbose=verbose)
    seq_opts = SeqOpts(subtype=subtype, keep_signal=False)
    aadiff_opts = AadiffOpts(
        make_consensus=make_consensus,
        consensus_as_reference=consensus_as_reference,
        remove_unchanged=False,
        annotation_tables=annotation_tables,
        join_annotations=join_annotations,
        caton82=caton82,
        wiley81=wiley81,
    )
    for row in fun.referenced_annotation_table(faa, mafft_opts, seq_opts, aadiff_opts):
        print("\t".join(row))


@click.command(
    name="represent",
    help="Representative subsampling by sequence similarity. The input must be a sequence alignment.",
)
@click.argument("alignment", default=sys.stdin, type=click.File())
@click.option(
    "--max-day-sep",
    help="Maximum number of days separating members of a group",
    type=click.IntRange(min=0, max=INT_SENTINEL),
    default=INT_SENTINEL,
)
@click.option(
    "--min-pident-sep",
    help="Minimum proportion identity difference between members of a group",
    type=click.FloatRange(min=0, max=1),
    default=1,
)
@click.option(
    "--same-state", help="Group strains only if they are in the same sate", is_flag=True
)
@click.option(
    "--print-groups",
    help="Rather than subsetting the fasta, print the groups of similar strains",
    is_flag=True,
)
def represent_cmd(
    alignment: TextIO,
    max_day_sep: Optional[int],
    min_pident_sep: float,
    same_state: bool,
    print_groups: bool,
):
    if max_day_sep == INT_SENTINEL:
        max_day_sep = None

    (groups, seqs) = fun.with_aligned_pairs(
        alignment,
        fun.represent,
        max_day_sep=max_day_sep,
        min_pident_sep=min_pident_sep,
        same_state=same_state,
    )

    if print_groups:
        for group in groups:
            for i in group:
                print(seqs[i][0])
            print("")
    else:
        for (header, seq) in seqs:
            print(">" + header)
            print(seq)


conversion_opt = click.option(
    "--conversion",
    type=click.Choice(["dna2aa", "dna2dna", "aa2aa"], case_sensitive=False),
    default="dna2aa",
    help="aa2aa: align the input and reference AA sequences and extract the HA1 regions based on the reference. dna2aa: translate the input DNA sequences, add the AA reference, align, and extract the HA1. dna2dna: translate the input DNA (keeping track of the CDS start position), add reference, align, find AA HA1 motif, map HA1 regions back to DNA.",
)


@click.command(name="ha1")
@click.argument("fasta_file", default=sys.stdin, type=click.File())
@mafft_exe_opt
@verbose_opt
@ha_subtype_no_keep_opt
@conversion_opt
def ha1_cmd(
    fasta_file: TextIO, mafft_exe: str, verbose: bool, subtype: str, conversion: str
):
    """
    Trim down to the HA1 using using subtype references from (Burke 2014). The
    HA1 bounds calibrated to reproduce the HA1 regions that are reported in
    genbank.
    """

    conversion_enum = parse_conversion(conversion)
    mafft_opts = MafftOpts(mafft_exe=mafft_exe, verbose=verbose)
    seq_opts = SeqOpts(subtype=subtype, conversion=conversion_enum)

    fun.print_fasta(
        fun.extract_ha1(
            fasta_file=fasta_file,
            mafft_opts=mafft_opts,
            seq_opts=seq_opts,
        )
    )


@click.command(
    name="motif",
)
@click.argument("fasta_file", default=sys.stdin, type=click.File())
@keep_signal_opt
@mafft_exe_opt
@verbose_opt
@as_fasta_opt
@motif_range_opt
@subtype_no_keep_opt
@conversion_opt
def motif_cmd(
    fasta_file: TextIO,
    keep_signal: bool,
    fasta: bool,
    mafft_exe: str,
    verbose: bool,
    motif: List[str],
    subtype: str,
    conversion: str,
):
    """
    Extract a motif based on indices relative to the reference (Burke 2014).

    Example:

      flutile trim motif --subtype=H1 -m "Cb=69-75" -m "Ca2=154-159,163-164"

    This would create a table with the fasta deflines in the first column and
    the Cb and Ca2 motifs in the next two columns.

    If no name is given, for example in `-m "69-79"`, the column name will
    default to the name of the subtype followed by the bounds ("H1:69-79").

    You can get the H3 antigenic motif as so:

       flutile trim motif --subtype=H3 -m "motif=145,155,156,158,159,189"

    If the --fasta flag is set, a fasta file with just the motifs is created
    instead of a table. This would allow, for example, the extraction of the
    HA2 region by reference index.
    """

    from flutile.api import write_bounds

    write_bounds(
        motif_strs=motif,
        subtype=subtype,
        fasta_file=fasta_file,
        tabular=not fasta,
        keep_signal=keep_signal,
        conversion=conversion,
        outfile=sys.stdout,
        mafft_exe=mafft_exe,
        verbose=verbose,
    )


@click.group(
    name="trim",
    help="Trim flu sequences in various ways. The trim operations use subtype-specific templates that are stored in the flutile package. You should not need to change these.",
    context_settings=CONTEXT_SETTINGS,
)
def trim_grp():
    pass


trim_grp.add_command(ha1_cmd)
trim_grp.add_command(motif_cmd)


@click.group(help="Flu-crew utilities", context_settings=CONTEXT_SETTINGS)
@click.version_option(__version__, "-v", "--version", message=__version__)
def cli_grp():
    pass


cli_grp.add_command(aadiff_cmd)
cli_grp.add_command(annotate_cmd)
cli_grp.add_command(represent_cmd)
cli_grp.add_command(trim_grp)


def main():
    cli_grp()
